module/zfs/vdev_raidz.c

```diff
@@ -3141,9 +3217,12 @@ raidz_reflow_impl(vdev_t *vd, vdev_raidz_expand_t *vre, range_tree_t *rt,
 	    ZIO_FLAG_CANFAIL,
 	    raidz_reflow_write_done, rra);
 
+	roffset = (blkid / old_children) << ashift;
+	if (vre->vre_scratch_devices != 0)
+		roffset -= VDEV_BOOT_SIZE;
```

Comment:
The problem that the scratch space is addressing is:
* we just started an expansion
* a single stripe has overwritten itself, such that it now has 2 sectors on the same disk
* that disk dies
* we now need to reconstruct this stripe, but we don't have enough of its sectors

To address this we need to ensure that when moving blocks, a stripe never overwrites itself, so that the stripe is always available intact at its old location.

So the place where we need to use the scratch copy is when doing a normal (or scrub/resilver) read.  We don't need to use the scratch copy when doing the reflow read (although it shouldn't hurt).

I don't think we can test this case currently, because we don't handle disk failure during reflow.  That's another thing on the to-do list.  We need to pause the reflow and wait for the disk to be replaced and resilvered.

However, I think we can test this by reading blocks that are in the beginning of the disk, while we are reflowing the beginning of the disk.  Today this will simply read from the old location, see this code in `vdev_raidz_map_alloc_expanded()`:

```c
		/*
		 * If we are in the middle of a reflow, and any part of this
		 * row has not been copied, then use the old location of
		 * this row.
		 */
		int row_phys_cols = physical_cols;
		if (b + (logical_cols - nparity) > reflow_offset >> ashift)
			row_phys_cols--;
```

I think that if the expansion has overwritten a stripe with itself, the read will not get the right data (because the old location has been partially overwritten).  We could test this by pausing the reflow in the very beginning, and then reading the block containing the partially-overwritten stripe.  I think that @stuartmaybee is working on some code for pausing the reflow for testing purposes.

Note that in this test case, we actually still have all the data and we could write code to read this "split stripe" partially from the old location and partially from the new location.  But that code doesn't exist, because it won't be needed once we solve the problem of handling resilvering while in the middle of reflow (by using the scratch copy).

Reply:
Oh, the reflow-pausing code is already there, `zfs_raidz_expand_max_offset_pause`